[Remote Store] Handoff refreshes, translog uploads during relocation from old to new primary

ashking94 opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
This PR fixes an edge case during primary relocation that can lead to both existing primary and new primary uploading segments/translog metadata having same combination of primary term, segment info generation, translog generation under following scenarios.
### Situations - 
1. Internal flush triggered during shard inactivity or snapshot just before the primary mode of the existing primary is set to false. This is race condition where the new primary starts itself with generation n, but n+1 generation gets uploaded to remote store by existing primary. 
2. New primary does a syncTranslog in [here](https://github.com/opensearch-project/OpenSearch/blob/5bb6caec906f9e89d330332ebb74789571409eb1/server/src/main/java/org/opensearch/indices/recovery/RecoveryTarget.java#L256-L259) before actually becoming a primary from cluster manager’s perspective. It is possible that the old primary can continue to be primary due to relocation handoff failure.

This can lead to following impact: 
#### situation no. 1  -
1. Red or Yellow index if the primary or replica shard goes unassigned due to any reasons (like node leaving cluster).
2. Primary relocation failure if there is no indexing happening.

The above can be mitigated by flushing multiple times if the index is not red. This can still be fixed by deleting the metadata file uploaded by old primary.

#### situation no. 2  -
1. In cases where the translog gets uploaded from new primary, but metadata files fails. If a node drop happens, remote store restore will restore a translog that has been overwritten by new primary

The above issue can go away if the indexing continues for some more time.

### Related Issues
Resolves #11320, #11322, #11323
<!-- List any other related issues here -->

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [ ] ~Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

ashking94 commented:
Need to validate that we are able to not reproduce the issue mentioned in https://github.com/opensearch-project/OpenSearch/pull/10839

sachinpkale commented on server/src/main/java/org/opensearch/index/translog/TranslogManager.java:
```diff
@@ -135,5 +136,10 @@ public interface TranslogManager {
+    /**
+     * Drains ongoing syncs to the underlying store. It returns a releasable which can be closed to resume the syncs back.
+     */
+    Releasable drainSyncToStore();
```
nit: Do we need `ToStore` here? It is implied in `sync`

sachinpkale commented on server/src/main/java/org/opensearch/index/translog/Translog.java:
```diff
@@ -1817,6 +1817,13 @@ protected void setMinSeqNoToKeep(long seqNo) {}
+    /**
+     * Drains ongoing syncs to the underlying store. It returns a releasable which can be closed to resume the syncs back.
+     */
+    protected Releasable drainSyncToStore() {
```
Will local translog use this implementation? In that case, a no-op may not be the expected behavior. Until we get the actual implementation added to local translog, can we throw `NotSupportedException`?

ashking94 replied:
Local translog will use the above no-op implementation. This code will be executed for non-remote store indexes as well. There are multiple test/ITs failure if we throw exception for local translog case. 

With the no-op implementation, we let the status quo behaviour be there for non-remote store cases. Otherwise we will need to have multiple if else block to run this only for remote indexes.

Bukhtawar commented on server/src/main/java/org/opensearch/index/shard/CloseableRetryableRefreshListener.java:
```diff
@@ -184,23 +187,38 @@ private void scheduleRetry(boolean afterRefreshSuccessful, boolean didRefresh) {
 
     protected abstract Logger getLogger();
 
+    // Made available for unit testing purpose only
+    /**
+     * Returns the timeout which is used while draining refreshes.
+     */
+    protected TimeValue getDrainTimeout() {
+        return DRAIN_TIMEOUT;
+    }
+
```
Make this package private?

Bukhtawar commented on server/src/main/java/org/opensearch/index/shard/CloseableRetryableRefreshListener.java:
```diff
@@ -26,14 +27,16 @@
  * is closed, all the permits are acquired and there are no available permits to afterRefresh. This abstract class provides
  * necessary abstract methods to schedule retry.
  */
-public abstract class CloseableRetryableRefreshListener implements ReferenceManager.RefreshListener, Closeable {
+public abstract class CloseableRetryableRefreshListener implements ReferenceManager.RefreshListener {
```
Want to rename this to `ReleasableRetryableRefreshListener`

Bukhtawar commented on server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:
```diff
@@ -316,13 +322,13 @@ private boolean prepareAndUpload(Long primaryTerm, Long generation) throws IOExc
+        // action which re-downloads the segments and translog on the new primary. We are ensuring 2 things here -
+        // 1. Using primaryModeSupplier, we prevent the new primary to do pre-emptive syncs
+        // 2. Using syncPermits, we prevent syncs at the desired time during primary relocation.
+        if (primaryModeSupplier.getAsBoolean() == false || syncPermits.tryAcquire(1) == false) {
```
nit: avoid hard coding use `SYNC_PERMITS`

Bukhtawar replied:
With these new permits we are strictly serialising the uploads, could it have been possible to have concurrent uploads in the present world?

ashking94 replied:
Concurrent uploads per shard is not possible today due to the write lock and read lock acquired during the sync. The lock sequencing ensures that there is only 1 upload any time.

gbbafna replied:
yes, it is possible for concurrent uploads case. However i am not able to find a better way to allow that in the current flow.

ashking94 replied:
@gbbafna How are you deducing that there are concurrent uploads cases? The `upload` method in RemoteFsTranslog is only called by another instance method `prepareAndUpload`. For control to enter inside the method, it needs the write lock which is later downgraded to read lock. Still, as the read lock is acquired, no one else can acquire the write lock so any other upload needs to wait until the current upload is over. 

Let me know if there is any case where we can upload concurrently.

Bukhtawar commented on server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:
```diff
@@ -423,11 +431,35 @@ protected void setMinSeqNoToKeep(long seqNo) {
     }
 
+    @Override
+    protected Releasable drainSync() {
+        try {
+            if (syncPermits.tryAcquire(SYNC_PERMITS, 1, TimeUnit.MINUTES)) {
+                logger.info("All permits acquired");
+                return Releasables.releaseOnce(() -> {
+                    syncPermits.release(SYNC_PERMITS);
+                    assert syncPermits.availablePermits() == SYNC_PERMITS : "Available permits is " + syncPermits.availablePermits();
+                    logger.info("All permits released");
+                });
+            } else {
+                throw new TimeoutException("Timeout while acquiring all permits");
+            }
+        } catch (TimeoutException | InterruptedException e) {
+            throw new RuntimeException("Failed to acquire all permits", e);
+        }
```
1. Move logger to debug or provide more context
2. Does 1 min align with SDK retries and timeouts, I doubt?

ashking94 replied:
For no 1, adding more context in the logs. For no 2, when this method is called during primary relocation, the translog upload has already finished for all indexing that has happened until the indexing block kicked in. This acquire is to ensure that if there is any ongoing internal flush (triggered by flushOnIdle, Snapshot or similar) which has no additional data, should finish before we start the force segrep step on the new primary. This has been set a bit lenient for general happy cases across all possible repository plugin implementation.

Bukhtawar commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -896,6 +903,15 @@ public void relocated(
             failShard("timed out waiting for relocation hand-off to complete", null);
             throw new IndexShardClosedException(shardId(), "timed out waiting for relocation hand-off to complete");
+        } finally {
+            // If the primary mode is still true after the end of handoff attempt, it basically means that the relocation
+            // failed. The existing primary will continue to be the primary, so we need to allow the segments and translog
+            // upload to resume.
+            if (replicationTracker.isPrimaryMode()) {
+                for (Releasable releasable : releasablesOnNoHandoff) {
+                    releasable.close();
+                }
+            }
```
Lets use explicit exception block to catch a relocation failure and add an assertion on primary mode

ashking94 replied:
Added an explicit catch block as suggested. fyi - The primaryMode is the indication for the final handoff status while exiting the below method -
https://github.com/opensearch-project/OpenSearch/blob/5bb6caec906f9e89d330332ebb74789571409eb1/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L848

Bukhtawar commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -896,6 +903,15 @@ public void relocated(
+            // failed. The existing primary will continue to be the primary, so we need to allow the segments and translog
+            // upload to resume.
+            if (replicationTracker.isPrimaryMode()) {
+                for (Releasable releasable : releasablesOnNoHandoff) {
+                    releasable.close();
+                }
```
You can use the utilitity `Releasables.close`

gbbafna replied:
Should we move this `release` part to `abortRelocationHandoff` ? This is a much better fit there .

ashking94 replied:
The abortRelocationHandoff might not run if the failure happens before the control reaches the below code -
https://github.com/opensearch-project/OpenSearch/blob/5bb6caec906f9e89d330332ebb74789571409eb1/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L877-L890

The failures can happen during draining refreshes or translog sync. With this being in finally, we ensure that the rollback step runs irrespective of where the failure occurred.

Bukhtawar commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -857,11 +860,15 @@ public void relocated(
+                }
+
+                // Ensure all in-flight remote store translog upload drains, before we perform the performSegRep.
+                releasablesOnNoHandoff.add(getEngineOrNull().translogManager().drainSync());
```
It will throw a NPE if engine returned is null, lets handle it explicitly

Bukhtawar commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -845,6 +845,9 @@ public void relocated(
         assert shardRouting.primary() : "only primaries can be marked as relocated: " + shardRouting;
+        // The below list of releasable ensures that if the relocation does not happen, we undo the activity of close and
+        // acquire all permits. This will ensure that the remote store uploads can still be done by the existing primary shard.
+        List<Releasable> releasablesOnNoHandoff = new ArrayList<>();
```
Initialise with size 2

gbbafna replied:
nit : `releasablesOnAbortHandoff`

gbbafna commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -3883,6 +3899,15 @@ public boolean isRemoteTranslogEnabled() {
+     * cluster state, the shard will not be in STARTED state. This method is used to prevent pre-emptive segment or
+     * translog uploads.
+     */
+    public boolean isStartedPrimary() {
+        return getReplicationTracker().isPrimaryMode() && state() == IndexShardState.STARTED;
+    }
```
will this affect the shards we are initializing from snapshot/remote store restore/local recovery ? As per the code , it shouldn't  . But would be good to double check this .

ashking94 replied:
have planned this already - https://github.com/opensearch-project/OpenSearch/pull/11330/#issuecomment-1826215993

gbbafna commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -3865,10 +3881,10 @@ private EngineConfig newEngineConfig(LongSupplier globalCheckpointSupplier) thro
             tombstoneDocSupplier(),
             isReadOnlyReplica,
-            replicationTracker::isPrimaryMode,
+            this::isStartedPrimary,
```
Lets rename the argument name from `primaryModeSupplier` to `isStartedPrimary` as well in `EngineConfigFactory` .

gbbafna commented on server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:
```diff
@@ -53,7 +57,6 @@
-    private final BlobStoreRepository blobStoreRepository;
     private final TranslogTransferManager translogTransferManager;
     private final FileTransferTracker fileTransferTracker;
     private final BooleanSupplier primaryModeSupplier;
```
nit : rename to `startedPrimarySupplier`.

ashking94 commented:
> ❌ Gradle check result for [bdb92d4](https://github.com/opensearch-project/OpenSearch/commit/bdb92d478dc83f7f448dd3e4dc76f845cbc5d58b): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/30386/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky tests - https://github.com/opensearch-project/OpenSearch/issues/10755, https://github.com/opensearch-project/OpenSearch/issues/11258

ashking94 commented:
> ❌ Gradle check result for [37f713a](https://github.com/opensearch-project/OpenSearch/commit/37f713a9f00ae8199c3d192057de77ed1fe126d2): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/30387/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

One flaky test https://github.com/opensearch-project/OpenSearch/issues/5329, and one failure in newly written test. Currently reproducing the failure. Will fix it.

ashking94 commented:
> ❌ Gradle check result for [954a43d](https://github.com/opensearch-project/OpenSearch/commit/954a43dbc5f521783632b05696b2d140520c3a3f): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/30388/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test - https://github.com/opensearch-project/OpenSearch/issues/10006

ashking94 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -3883,6 +3901,15 @@ public boolean isRemoteTranslogEnabled() {
+     * cluster state, the shard will not be in STARTED state. This method is used to prevent pre-emptive segment or
+     * translog uploads.
+     */
+    public boolean isStartedPrimary() {
```
The same check is now used for translog and segments upload both.

ashking94 commented on server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:
```diff
@@ -503,11 +502,11 @@ private boolean isReadyForUpload() {
                 sb.append(" engineType=").append(indexShard.getEngine().getClass().getSimpleName());
             }
-            if (isLocalOrSnapshotRecovery() == false) {
+            if (indexShard.recoveryState() != null) {
```
isLocalOrSnapshotRecovery() checks for recoveryState. recoveryState can be null in which case this check throw NPE. Hence adding this check.

Bukhtawar replied:
You can always re-add `isLocalOrSnapshotRecovery` after a null check?

ashking94 replied:
yes, but we need to log the variables regardless of the final return value.

ashking94 commented on server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:
```diff
@@ -341,10 +348,17 @@ private boolean upload(long primaryTerm, long generation, long maxSeqNo) throws
                 new RemoteFsTranslogTransferListener(generation, primaryTerm, maxSeqNo)
             );
+        } finally {
+            syncPermit.release(SYNC_PERMIT);
```
This is deterministic to run if `syncPermit.tryAcquire(SYNC_PERMIT) == false` does not hold true.

ashking94 commented on server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:
```diff
@@ -341,10 +348,17 @@ private boolean upload(long primaryTerm, long generation, long maxSeqNo) throws
     }
 
+    // Made available for testing only
+    void induceDelayForTest() {
```
merely for testing a case which otherwise is difficult to test.

Bukhtawar replied:
The way to do this is to use an overridden(mock) version of `RemoteFsTranslog` that induces the delay before the base method invocation

ashking94 commented on server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:
```diff
@@ -505,11 +543,7 @@ public static void cleanup(Repository repository, ShardId shardId, ThreadPool th
     }
 
     protected void onDelete() {
-        if (primaryModeSupplier.getAsBoolean() == false) {
```
The delete is invoked from IndexShard where the primaryMode check is present already.

https://github.com/opensearch-project/OpenSearch/blob/5bb6caec906f9e89d330332ebb74789571409eb1/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L1989-L1992

This code is only triggered in case of index deletion in which case we should let the deletion complete.

Here, we have added to assert that this can only be called by cluster applier thread. Will add a comment as well.

Bukhtawar commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -896,6 +903,17 @@ public void relocated(
+            assert replicationTracker.isPrimaryMode();
+            throw ex;
+        } finally {
+            // If the primary mode is still true after the end of handoff attempt, it basically means that the relocation
+            // failed. The existing primary will continue to be the primary, so we need to allow the segments and translog
+            // upload to resume.
+            if (replicationTracker.isPrimaryMode()) {
+                Releasables.close(releasablesOnHandoffFailures);
```
why can't this be moved to the catch block above?

Bukhtawar commented on server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:
```diff
@@ -503,11 +502,11 @@ private boolean isReadyForUpload() {
                 sb.append(" primary=").append(indexShard.shardRouting.primary());
             }
-            logger.trace(sb.toString());
+            logger.info(sb.toString());
```
Why info?

ashking94 replied:
This is helpful in debugging cases where the refresh listener's refresh method is invoked but it is not ready for upload due to any reason.

ashking94 replied:
There will be very few instances when this will be logged. for eg - create index, failover, relocation. This will not bombard the logs.

Bukhtawar commented on server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:
```diff
@@ -423,11 +437,35 @@ protected void setMinSeqNoToKeep(long seqNo) {
         // clean up local translog files and updates readers
         super.trimUnreferencedReaders();
 
+        // This is to ensure that after the permits are acquired during primary relocation, there are no further modification on remote
+        // store.
+        if (syncPermit.availablePermits() != SYNC_PERMIT) {
+            return;
+        }
+
```
This can also impact other flows while like when a concurrent upload is happening

ashking94 replied:
Used the same approach used in ReleasableRefreshListener. This would not be impacted by concurrent upload now.

Bukhtawar commented on server/src/main/java/org/opensearch/index/translog/Translog.java:
```diff
@@ -1817,6 +1817,13 @@ protected void setMinSeqNoToKeep(long seqNo) {}
     protected void onDelete() {}
 
+    /**
+     * Drains ongoing syncs to the underlying store. It returns a releasable which can be closed to resume the syncs back.
+     */
+    protected Releasable drainSync() {
+        return () -> {}; // noop
+    }
+
```
Will prefer this abstract and make LocalTranslog noop instead so that new implementors explicitly override this

gbbafna commented on server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:
```diff
@@ -489,8 +489,7 @@ private void initializeRemoteDirectoryOnTermUpdate() throws IOException {
      * @return true iff primaryMode is true and index shard is not in closed state.
```
modify the java. doc as well

gbbafna commented on server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:
```diff
@@ -516,8 +515,8 @@ private boolean isLocalOrSnapshotRecovery() {
         // all segments to remote before completing the recovery to ensure durability.
-
         return (indexShard.state() == IndexShardState.RECOVERING && indexShard.shardRouting.primary())
+            && indexShard.recoveryState() != null
```
`recoveryState` is never null . It is set to `DONE` after the recovery is completed  .

ashking94 replied:
The recoveryState is nullable and is not set in the ctor as well. There is a possibility that recoveryState can be null.

https://github.com/opensearch-project/OpenSearch/blob/74b2d7daf4f5caf42d71724b2dcb745a31e6a884/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L297-L298

https://github.com/opensearch-project/OpenSearch/blob/74b2d7daf4f5caf42d71724b2dcb745a31e6a884/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L2539-L2546

Let me know if you still think it can never be null.

gbbafna commented on server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:
```diff
@@ -316,17 +322,18 @@ private boolean prepareAndUpload(Long primaryTerm, Long generation) throws IOExc
     }
 
     private boolean upload(long primaryTerm, long generation, long maxSeqNo) throws IOException {
-        // During primary relocation (primary-primary peer recovery), both the old and the new primary have engine
-        // created with the RemoteFsTranslog. Both primaries are equipped to upload the translogs. The primary mode check
-        // below ensures that the real primary only is uploading. Before the primary mode is set as true for the new
-        // primary, the engine is reset to InternalEngine which also initialises the RemoteFsTranslog which in turns
-        // downloads all the translogs from remote store and does a flush before the relocation finishes.
-        if (primaryModeSupplier.getAsBoolean() == false) {
-            logger.debug("skipped uploading translog for {} {}", primaryTerm, generation);
+        // During primary relocation, both the old and new primary have engine created with RemoteFsTranslog and having
+        // ReplicationTracker.primaryMode() as true. However, before we perform the `internal:index/shard/replication/segments_sync`
+        // action which re-downloads the segments and translog on the new primary. We are ensuring 2 things here -
+        // 1. Using startedPrimarySupplier, we prevent the new primary to do pre-emptive syncs
+        // 2. Using syncPermits, we prevent syncs at the desired time during primary relocation.
+        if (startedPrimarySupplier.getAsBoolean() == false || syncPermit.tryAcquire(SYNC_PERMIT) == false) {
+            logger.debug("skipped uploading translog for {} {} syncPermits={}", primaryTerm, generation, syncPermit.availablePermits());
             // NO-OP
             return true;
         }
```
should we move this to `prepareAndUpload` ? There is no point creating another generation locally , if we can't rotate it .

ashking94 replied:
Do you mean to say `upload` instead of `roatate`?

In context of upload, I will check this a bit more if we can move this check to prepareAndUpload and I think we might have discussed this in the older PR. Also, we might need similar change for segments i.e. prevent segments or segment infos creation if we can't upload.

gbbafna replied:
This would be tricky for segments as the creation of local segments in not in our control as of now , unlike translog .

ashking94 commented:
> ❌ Gradle check result for [2efefd9](https://github.com/opensearch-project/OpenSearch/commit/2efefd97bd933942ef8f443eecce1b67e305c355): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/30405/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test - https://github.com/opensearch-project/OpenSearch/issues/10704 with same stack trace as earlier.

ashking94 commented:
> ❌ Gradle check result for [a15e6d6](https://github.com/opensearch-project/OpenSearch/commit/a15e6d6bb564fa7b48c96e193eba2f8c9b8490b5): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/30406/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

flaky test - https://github.com/opensearch-project/OpenSearch/issues/10785 (same assertion failure as reported earlier)

gbbafna commented:
LGTM. Thanks for the changes @ashking94 . Would be merge ready post resolving outstanding comments/issue.

ashking94 commented:
> LGTM. Thanks for the changes @ashking94 . Would be merge ready post resolving outstanding comments/issue.

@gbbafna  The below are the open comments - 
Open comments - https://github.com/opensearch-project/OpenSearch/pull/11330#discussion_r1405429203, https://github.com/opensearch-project/OpenSearch/pull/11330#discussion_r1405181625

@Bukhtawar @sachinpkale  I have addressed your comments. Please take a look.

ashking94 commented:
> > LGTM. Thanks for the changes @ashking94 . Would be merge ready post resolving outstanding comments/issue.
> 
> @gbbafna The below are the open comments - Open comments - [#11330 (comment)](https://github.com/opensearch-project/OpenSearch/pull/11330#discussion_r1405429203), [#11330 (comment)](https://github.com/opensearch-project/OpenSearch/pull/11330#discussion_r1405181625)
> 
> @Bukhtawar @sachinpkale I have addressed your comments. Please take a look.

I have resolved all the comments.

sachinpkale commented on server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:
```diff
@@ -423,11 +433,39 @@ protected void setMinSeqNoToKeep(long seqNo) {
 
+        // This is to ensure that after the permits are acquired during primary relocation, there are no further modification on remote
+        // store.
+        if (pauseSync.get()) {
```
Thinking if we need a check if primary mode started here as well

ashking94 replied:
Made the change.

ashking94 commented:
> ❌ Gradle check result for [61a7fc3](https://github.com/opensearch-project/OpenSearch/commit/61a7fc31a19f6ecfff620a103e56efd66a13cb9c): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/30424/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Known and existing flaky test - https://github.com/opensearch-project/OpenSearch/issues/6542

ashking94 commented:
> testDelayWithALargeAmountOfShards

Flaky test - https://github.com/opensearch-project/OpenSearch/issues/10558 with same error.

Bukhtawar commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -896,6 +903,14 @@ public void relocated(
             failShard("timed out waiting for relocation hand-off to complete", null);
             throw new IndexShardClosedException(shardId(), "timed out waiting for relocation hand-off to complete");
+        } catch (Exception ex) {
+            logger.warn("exception occurred during relocation hand-off to complete errorMsg={}", ex.getMessage());
```
logger.error("Exception occurred during relocation hand-off to complete", ex);

ashking94 replied:
Removing this log line as we are throwing the same exception as is. I have verified that the original exception shows up from the logger of `IndicesClusterStateService`. Having same log with complete stack trace twice will only litter the code and logs.

ashking94 replied:
resolving this.

ashking94 commented:
> ❕ Gradle check result for [ea629f4](https://github.com/opensearch-project/OpenSearch/commit/ea629f41c25e0bf5015a23594cd24b169b4690e6): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/30490/)
> 
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.remotestore.RemoteIndexPrimaryRelocationIT.testPrimaryRelocationWhileIndexing
>       1 org.opensearch.cluster.allocation.ClusterRerouteIT.testDelayWithALargeAmountOfShards
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

Flaky test - https://github.com/opensearch-project/OpenSearch/issues/10558

Bukhtawar commented on server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:
```diff
@@ -267,6 +275,16 @@ public void rollGeneration() throws IOException {
+        // action which re-downloads the segments and translog on the new primary. We are ensuring 2 things here -
+        // 1. Using startedPrimarySupplier, we prevent the new primary to do pre-emptive syncs
+        // 2. Using syncPermits, we prevent syncs at the desired time during primary relocation.
+        if (startedPrimarySupplier.getAsBoolean() == false || syncPermit.tryAcquire(SYNC_PERMIT) == false) {
+            logger.debug("skipped uploading translog for {} {} syncPermits={}", primaryTerm, generation, syncPermit.availablePermits());
+            // NO-OP
+            return false;
+        }
```
Why can't we simply use `ReleasableLock#tryAcquire` to make it consistent with read/write locks instead of introducing new permits?

ashking94 replied:
Will be evaluating other sempahores as well and replace with ReleasableLock. Created issue - https://github.com/opensearch-project/OpenSearch/issues/11360

ashking94 commented:
Raising manual backport since the automated one failed.
